### PR TITLE
fix(core/anchor-expander): clone nodes when expanding an anchor

### LINKED ISSUE #2131: Linking to sections & figures
This has become so annoying that i'm finally making time to raise an issue. Currently automatic completion of link text for links pointing to sections and figures is prompted by creating a `a` element without link text, eg. `<a href="#myfragid"></a>`. This produces no visible rendered text if the link fails, and also in Dreamweaver's design view (which i use all the time). You get something such as:

> See .

This is annoying because i can't see at a glance where it's pointing to. If the rendering fails, it's often not very noticeable, and you have to jump through some hoops to locate the appropriate bit of source text in order to know which fragid is failing.

Could we create an alternative (to allow backwards compatibility) where the souce code would look like this:

```
<a class="secref">myfragid</a>
```

This would then show the following in design view, or in failed renderings:

> See myfragid

For figure references we could use `<a class="figref">myfragid</a>`.

I use this approach for my own document templates, and it works well.  The essential logic shouldn't change much, just get the cue from the class name and the target location from the link text.

### PR DESCRIPTION
**Pull request by marcoscaceres**
closes #2131

Refactors expanding anchors into its own module. This now supports safely cloning a referenced node (by removing ids and converting `dfn`s to spans), and applies localization heuristics, so the correct language and directionality are applied to an expanded reference.

### CODE REVIEW

File: src/core/figures.js [hunk 1]
```diff
@@ -48,23 +46,20 @@ export function run(conf) {
...
       decorateFigure(fig, caption, i, conf);
-      figMap[fig.id] = caption.childNodes;
     } else {
       showInlineWarning(fig, "Found a `<figure>` without a `<figcaption>`");
     }
 
     tof.push(getTableOfFiguresListItem(fig.id, caption));
   });
-  return { figMap, tof };
+  return { tof };
```
Comment (saschanaz) [inline src/core/figures.js]:
```suggestion
  return tof;
```

File: src/core/inlines.js [hunk 2]
```diff
@@ -50,10 +50,10 @@ function inlineRefMatches(matched) {
   // slices "[[[" at the beginning and "]]]" at the end
   const ref = matched.slice(3, -3).trim();
   if (!ref.startsWith("#")) {
-    return hyperHTML`<a data-cite="${ref}"></a>`;
+    return hyperHTML`<a class="respec-inline-expansion" data-cite="${ref}"></a>`;
```
Comment (saschanaz) [inline src/core/inlines.js]: Is this only for testing?

File: src/core/structure.js [hunk 3]
```diff
@@ -88,13 +85,12 @@ function scanSections(sections, maxTocLevel, { prefix = "" } = {}) {
...
       });
       if (sub) {
-        Object.assign(secMap, sub.secMap);
         item.append(sub.ol);
       }
       ol.append(item);
     }
   }
-  return { ol, secMap };
+  return { ol };
```
Comment (saschanaz) [inline src/core/structure.js]:
```suggestion
  return ol;
```

File: src/core/anchor-expander.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+function makeSafeCopy(node) {
+  const clone = node.cloneNode(true);
+  clone.querySelectorAll("[id]").forEach(elem => elem.removeAttribute("id"));
+  clone.querySelectorAll("dfn").forEach(dfn => {
+    const span = document.createElement("span");
+    span.append(...dfn.childNodes);
+    [...dfn.attributes].forEach(({ name, value }) => {
+      span.setAttribute(name, value);
+    });
+    dfn.parentNode.replaceChild(span, dfn);
```
Comment (saschanaz) [inline src/core/anchor-expander.js]: The whole thing in this loop could be replaced with `renameElement()` from `core/utils.js`.

File: src/core/anchor-expander.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+function localize(matchingElement, newElement) {
+  for (const attrName of ["dir", "lang"]) {
+    // Already set on element, don't override.
+    if (newElement.hasAttribute(attrName)) continue;
+    const matchingClosest = matchingElement.closest(`[${attrName}]`);
+    const newClosest = newElement.closest(`[${attrName}]`);
+
+    // It's the same, so already inherited from closet (probably HTML element or body).
+    if (matchingClosest === newClosest) continue;
```
Comment (saschanaz) [inline src/core/anchor-expander.js]: Shouldn't this be value comparison? Could get the same lang value even with different parents.

File: src/core/anchor-expander.js (hunk 4 above)
Comment (marcoscaceres) [reply src/core/anchor-expander.js]: oh! Forgot about that! thanks!!!

File: src/core/inlines.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/inlines.js]: I was thinking of styling these... I was going to add the "[[[" "]]]" back in using CSS generated content.

Comment (marcoscaceres) [reply src/core/inlines.js]: (styled when they fail, that is)

File: src/core/anchor-expander.js (hunk 5 above)
Comment (marcoscaceres) [reply src/core/anchor-expander.js]: yes, you are right. Well spotted.

File: js/profile-w3c-common.js [hunk 6]
```diff
@@ -79,6 +79,7 @@ define([
   "./core/highlight-vars",
   "./core/data-type",
   "./core/algorithms",
+  "./core/anchor-expander.js",
```
Comment (saschanaz) [inline js/profile-w3c-common.js]: Does this work? AFAIK requirejs refuses to work with `.js` file extension for relative paths.

Comment (marcoscaceres) [reply js/profile-w3c-common.js]: no, this will break :) I was testing only against w3c-respec profile. Fixing.

File: src/core/figures.js [hunk 7]
```diff
@@ -12,33 +12,46 @@ export const name = "core/figures";
 
 const localizationStrings = {
   en: {
-    table_of_fig: "Table of Figures",
+    list_of_figures: "List of Figures",
+    fig: "Figure ",
+  },
+  ja: {
+    fig: "図",
+    list_of_figures: "図のリスト ",
```
Comment (saschanaz) [inline src/core/figures.js]:
```suggestion
    list_of_figures: "図のリスト",
```

**Comment by marcoscaceres** [comment]
Thanks for fixing those @saschanaz!

File: src/core/inlines.js (hunk 2 above)
Comment (marcoscaceres) [reply src/core/inlines.js]: @saschanaz, decided to ditch the "respec-inline-expansion" class.

**Comment by marcoscaceres** [comment]
Ok, I think this might be good to go. I'll keep working with @r12a to improve the l10n aspects, but it covers the use cases that @r12a had.

**Pull request merged**